Update to an FCS that understands F# 7

baronfel opened the pull request:
Notable Changes:

* Our workarounds module could be deleted
* the type suffix for arrays was changed from `[]` to `array`
* SRTP type annotations were changed from  `^` to `'`
* Several codefixes (and tests) needed to be (temporarily) removed due to upstream issues that have been logged.

Test failure categories:
* [ ] We also seem to have an issue in completions where the full set of known names is being reported, and the namespace-probing isn't happening (e.g. so `Regex` isn't recognized and System.Text isn't offered as a suggested open). - punting to followup.
* [x] ! codefixes aren't returning (same with many others)
* [x] interpolation codefixes aren't finding an AST - anonymous modules aren't returning a good parse tree? - issue raised and resolved upstream already
* [x] Implement interface codefix tests are broken because of upstream issue (logged already), but the functionality mostly works - issue raised

baronfel committed:
WIP F# 7 update

This removes the 6.0 target while we wait for updated packages on NuGet.

baronfel commented on .vscode/launch.json:
```diff
@@ -18,32 +18,6 @@
       "stopAtEntry": false,
       "internalConsoleOptions": "openOnSessionStart"
     },
-    {
```
this is a duplicate launch config, so I removed it.

baronfel commented on global.json:
```diff
@@ -1,6 +1,5 @@
-        "rollForward": "latestPatch"
-    }
-}
+  "sdk": {
```
TODO: add this back when we get a 6.0-compatible package.

baronfel commented on src/FsAutoComplete.Core/CodeGeneration.fs:
```diff
@@ -952,7 +952,15 @@ module CodeGenerationUtils =
+    | SynType.Tuple (_, ts, _) ->
+      Some(
+        ts
+        |> Seq.choose (function
+          | SynTupleTypeSegment.Type name -> (|TypeIdent|_|) name
+          | SynTupleTypeSegment.Star _ -> Some "*"
+          | SynTupleTypeSegment.Slash _ -> Some "/")
```
There's now a concept of 'segments' of a tuple type declaration, so we piece them together bit by bit here.

baronfel commented on src/FsAutoComplete.Core/Commands.fs:
```diff
@@ -1156,7 +1156,7 @@ type Commands(checker: FSharpCompilerServiceChecker, state: State, hasAnalyzers:
 
   let fillHelpTextInTheBackground decls (pos: Position) fn getLine =
-    let declName (d: DeclarationListItem) = d.Name
+    let declName (d: DeclarationListItem) = d.NameInList
```
I chose to use `NameInList` everywhere in this PR, because it matches the semantic of `Name` in previous versions. We may want to use the other `NameInXX` property in specific scenarios, but that would be an enhancement later on.

kant2002 replied:
Great, I do the same choice, but because error/warning message was confusing I was not sure what to do. That's probably information more important to FSharp team, but anyway.

baronfel commented on src/FsAutoComplete.Core/DocumentationFormatter.fs:
```diff
@@ -131,8 +130,8 @@ module DocumentationFormatter =
           | _, _ ->
-            if PrettyNaming.IsMangledOpName c.MemberName then
-              PrettyNaming.DecompileOpName c.MemberName, false
+            if PrettyNaming.IsLogicalOpName c.MemberName then
+              PrettyNaming.ConvertValLogicalNameToDisplayNameCore c.MemberName, false
```
These two functions got a rename, nothing else.

baronfel commented on src/FsAutoComplete.Core/FCSPatches.fs:
```diff
@@ -380,6 +380,7 @@ module SyntaxTreeOps =
       | SynExpr.Null _
       | SynExpr.Ident _
       | SynExpr.ImplicitZero _
+      | SynExpr.Typar _
```
we have two new syntax nodes - Type Arguments, and Dynamic access. Both are relatively easy to adapt to.

baronfel commented on src/FsAutoComplete.Core/UntypedAstUtils.fs:
```diff
@@ -115,6 +116,9 @@ let internal getRangesAtPosition input (r: Position) : Range list =
       addIfInside r
       List.iter walkType ts
       walkMemberSig sign
+    | SynTypeConstraint.WhereSelfConstrained (t, r) ->
```
This is a new F# 7 feature - self constrained types. Walking them means checking the range and walking the contained type.

baronfel commented on src/FsAutoComplete.Core/UntypedAstUtils.fs:
```diff
@@ -521,6 +541,10 @@ let internal getRangesAtPosition input (r: Position) : Range list =
       Option.iter walkType t
       walkExpr e
     | SynMemberDefn.Open (longId, r) -> addIfInside r
+    | SynMemberDefn.GetSetMember (memberDefnForGet = getter; memberDefnForSet = setter; range = range) ->
```
There's a new wrapper for get/set members, easy to walk.

baronfel commented on src/FsAutoComplete.Core/Workaround/ServiceParseTreeWalk.fs:
```diff
@@ -1,1012 +0,0 @@
-// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
```
Deleted this file because I believe the workaround for match! is incorporated now.

baronfel commented on test/FsAutoComplete.Tests.Lsp/CompletionTests.fs:
```diff
@@ -497,10 +497,11 @@ let autoOpenTests state =
       | Ok (Some res) ->
         Expect.isFalse res.IsIncomplete "Result is incomplete"
-        let ci = res.Items |> Array.find (fun c -> c.Label = word)
+        let ci = res.Items |> Array.tryFind (fun c -> c.Label = word)
```
debugging here to try and figure out why we can't get all the completions.

baronfel commented on test/FsAutoComplete.Tests.Lsp/CoreTests.fs:
```diff
@@ -330,9 +330,9 @@ let tooltipTests state =
-                "   x: ^a (requires static member ( + ) ) ->"
-                "   y: ^b (requires static member ( + ) )"
-                "   -> ^c" ])
+                "   x: 'a (requires static member ( + ) ) ->"
```
hat types (^) are no longer required, so we should standardize onto normal backticks.

baronfel commented on test/FsAutoComplete.Tests.Lsp/ExtensionsTests.fs:
```diff
@@ -406,6 +406,6 @@ let signatureTests state =
     [ testList
         "tests"
-        [ verifySignature 0 (4, 16) "val arrayOfTuples: (int * int)[]"
+        [ verifySignature 0 (4, 16) "val arrayOfTuples: (int * int) array"
```
the default output for array types has changed from `[]` to `array` like other container types.

kant2002 commented on src/FsAutoComplete.Core/FsAutoComplete.Core.fsproj:
```diff
@@ -9,7 +9,6 @@
     <TrimmerRootAssembly Include="FsAutoComplete.Core" />
   </ItemGroup>
   <ItemGroup>
-    <Compile Include="Workaround/ServiceParseTreeWalk.fs" />
```
That removal was not obvious for me when I try to make things working.

edgarfgp commented:
@baronfel FSharp.Core 7 will now print `array` instead of `[]`. Does  FsAutoComplete some specific logic that needs to be removed?

baronfel commented:
The current blocked test is `FSAC.lsp.Ionide WorkspaceLoader.FSharpLspServer.Rename Tests.Across projects.Rename from usage across projects` - it's failing to load `TestCases\\RenameTest\\CrossProject\\LibB\\LibB.fsproj` successfully.

baronfel commented:
@edgarfgp that change has already been taken into account in this PR - the default formatting changed and so I updated the tests' expectations to match 👍

baronfel committed:
fix NuGet.Frameworks delay-load

baronfel commented:
Status: 

```
Failed:    2
Errored:  89
```

The failures and errors are mostly in the autocomplete namespace suggestions tests, so once that's fixed it should just be a bit of whack-a-mole.

baronfel committed:
fix a bunch of autocomplete tests

nojaf commented:
Hi @baronfel, could you use any help with this?
Is there anything blocking here? Like waiting for a dependency to update?
Or is the remaining work about fixing those last failing tests?

vzarytovskii commented:
> Hi @baronfel, could you use any help with this?
> Is there anything blocking here? Like waiting for a dependency to update?
> Or is the remaining work about fixing those last failing tests?

I would imagine it waits for new FCS update, the one without msbuild dependency. Should be out same time with 7.0.200

nojaf commented:
Would there be any reason not to use `43.7.200-preview.22569.1` if the MSBuild dependency thing indeed blocks everything? 
I imagine all the dependencies would also need to be bumped to that preview version.

nojaf commented:
Sorry, that was a bad suggestion, I just noticed that the preview FSC has a dependency on an `FSharp.Core` version that is not on NuGet:

![image](https://user-images.githubusercontent.com/2621499/213457841-194246c2-7a61-486a-9cc7-4ed933e32775.png)

baronfel committed:
workaround net7 test framework binding to nuget.frameworks

baronfel committed:
Fixup a few packaging errors after rebase

baronfel committed:
supply an actually working fcsCompletionContext that disappeared after rebase

baronfel committed:
default to adaptive LSP server

baronfel committed:
change ordering of implemented interface property getter/setter in test expectation to match reality

baronfel committed:
adjust to new FCS AST for property get/set on interface generation

also disable the 'without type annotation' interface generation entirely due to https://github.com/dotnet/fsharp/issues/14698

baronfel committed:
silence failing tests with a marker to reenable later

baronfel committed:
fix formatting

baronfel committed:
update to new stable FCS

baronfel committed:
fix c# lambda-to-fun codefix after the AST changed for operations

baronfel committed:
fix bang-to-value dereference codefix to account for operation names being LongIdents instead of Idents in the AST now

baronfel committed:
formatting

baronfel commented on test/FsAutoComplete.Tests.Lsp/FsAutoComplete.Tests.Lsp.fsproj:
```diff
@@ -42,5 +41,12 @@
+    <Copy SourceFiles="$([System.IO.Directory]::GetParent($(BundledRuntimeIdentifierGraphFile)))\NuGet.Frameworks.dll"
+          DestinationFolder="$(OutputPath)"
+          ContinueOnError="false" />
+  </Target>
```
I need to comment why this is needed.  short version here is that the `dotnet test` framework needs _a_ Nuget.Frameworks.dll, but in general we want to bind to the one from the SDK (because like MSBuild it needs to be runtime-bound).

baronfel committed:
Fix a few build issues and comment things

baronfel committed:
make the completion autoopen tests ignored for now

baronfel commented:
There's one more test failure around the new ^-formatting for types, but other than that I think we might be good to go.

baronfel committed:
ensure we no longer render hat-types at all in signature rendering
